initial revision of schedule.json export

mblaettler opened the pull request:
The goal of this commit is to export our schedule in a format that it is readable by the official "Fahrplan App". A working example is found here: https://pretalx.c3voc.de/fire-shonks-2022/schedule/export/schedule.json

The following topics need to be discussed/fixes:
- Is there a better way to manage timestamps (e.g. days.day_start or days[n].rooms.Winterchaos[n].date)?
- Is there a better way to handle the sessions guid (e.g. days[n].rooms.Winterchaos[n].guid)?
- Shoudl we intorduce different tracks (e.g. days[n].rooms.Winterchaos[n].track)?
- Should we export the speak duration to the session definition?
- Is the person.id and person.code calculated in a good way?

Commit:
initial revision of schedule.json export

The goal of this commit is to export our schedule in a format that it is readable
by the official "Fahrplan App". A working example is found here:
https://pretalx.c3voc.de/fire-shonks-2022/schedule/export/schedule.json
The following topics need to be discussed/fixes:
- Is there a better way to manage timestamps (e.g. days.day_start or days[n].rooms.Winterchaos[n].date)?
- Should we export the speak duration to the session definition?
- Is the person.id and person.code calculated in a good way?

gandro commented on _pages/schedule.json.liquid:
```diff
@@ -0,0 +1,88 @@
+        {
+          "index": {{ i }},
+          "date": "{{ day.name }}",
+          "day_start": "{{ day.name }}T00:00:00+01:00",
+          "day_end": "{{ day.name }}T00:00:00+01:00",
```
```suggestion
          "day_start": "{{ day.name }}T13:00:00+01:00",
          "day_end": "{{ day.name }}T23:59:59+01:00",
```

Otherwise the day will be empty I think.

gandro commented on _pages/schedule.json.liquid:
```diff
@@ -0,0 +1,88 @@
+            ]
+          }
+        }{% unless forloop.last %},{% endunless %}
+        {% assign i = i | plus:1 %}
```
Alternative:

```suggestion
        {% increment i %}
```

gandro commented on _pages/schedule.json.liquid:
```diff
@@ -0,0 +1,88 @@
+      {% for day in days %}
+        {
+          "index": {{ i }},
+          "date": "{{ day.name }}",
```
Not sure about the schema. Maybe this needs to be a fully qualified timestamp?

```suggestion
          "date": "{{ day.name }}T00:00:00+01:00",
```

gandro replied:
Edit: According to the validator it seems valid as is.

gandro commented on _pages/schedule.json.liquid:
```diff
@@ -0,0 +1,88 @@
+              {% for session in sessions %}
+              {
+                "id": "{{ session.talk_id }}",
+                "guid": "8e3616f7-df59-4f67-847d-783fea4a{{ session.talk_id }}",
```
Regarding the open question: Using GUID4 here is fine, I think. Let's do it this way.

Not sure if it was obvious, but the the talk id is DDHH (DD = day of the month, HH = hour of the day). So if a talk starts Dec 28 at 16:00, its talk_id is 2816. I rely on this for sorting the talks in the collection.

This means that if we move a talk, this guid will change. Which I think is fine? If that's not okay, then we should just store a unique random generated GUID in each talk file.

Kunsi replied:
Talk GUIDs *should* stay the same for each talk, so we (as in, the VOC pipeline) can accurately detect wether talks are new or just changed. If possible, we'd prefer to have static GUIDs.

gandro replied:
Thanks for the heads up! We will introduce static IDs then

gandro commented on _pages/schedule.json.liquid:
```diff
@@ -0,0 +1,88 @@
+    "days": [
+      {% assign days = site.sessions | group_by: "day" | sort: "name" %}
+      {% assign all_sessions = days | map: "items" %}
+      {% assign all_speakers = all_sessions | map: "speaker" | join: ", " | split: ", " | uniq %}
```
This is very clever :star_struck:

gandro reviewed (approved):
This is amazing work, thanks so much :rocket: 

I think we should merge this and try it out. Regarding the open questions:
 - I think the timestamp management is fine as is. We can always make it more dynamic if needed.
 - Session GUID is also fine with me. I think an alternative is to statically assign a random GUID to each talk in its yamlf ile. I left an inline comment on this.
 - Session tracks: I considered adding it to the schedule as well, so it would be a session variable. We can consider this for the future I think, but a single track is fine for now.
 - Speaking duration can be added later easily. The default 45mins is good enough for now.
 - Regarding person.{id,code}: I think with the current setup that's as good as we can get. Since the ID needs to be an integer, I don't have any better idea. Manually assigning it in the session definition is going to be annoying too.

gandro commented on _pages/schedule.json.liquid:
```diff
@@ -0,0 +1,88 @@
+        "capacity": null
+      }
+    ],
+    "days": [
```
I quickly used the https://www.jsonschemavalidator.net/ on this schema I found: https://github.com/voc/schedule/blob/198e253cc2d673e7a29945c55ba9dd1e8be24935/validator/json/schema.json

It seems this `days` attribute needs to be a child of  `conference` above. Other than that, it did not find any errors :partying_face:

gandro committed:
Make schedule pass JSON schema validation:

Two changes:

   - Move .days to .conference.days
   - Set day_start to 13:00 and day_end to 23:59

Signed-off-by: Sebastian Wicki <gandro@gmx.net>

gandro commented:
I've pushed some minor changes. The branch now passes JSON schema validation.
